README.md
[![Build Status](https://travis-ci.org/thehelp/messaging.svg?branch=master)](https://travis-ci.org/thehelp/messaging)

# thehelp-messaging

A lightweight package for sending SMS via [Twilio](https://www.twilio.com), and email via [Sendgrid](http://sendgrid.com/). Also makes it easy to receive SMS and email from these services in [`express`](http://expressjs.com/)-based apps.


## Setup

First, install the project as a dependency:

```bash
npm install thehelp-messaging --save
```

Then you'll need to provide your credentials for these services. You can do it in code, but I prefer to set it up with environment variables...

### Email via Sendgrid

[Create a Sendgrid sub-user account](https://sendgrid.com/credentials) and use its credentials:

```json
{
  "THEHELP_SENDGRID_USERNAME": "username",
  "THEHELP_SENDGRID_PASSWORD": "password"
}
```

### SMS via Twilio

Grab your Twilio API keys from your [account settings page](https://www.twilio.com/user/account/settings):

```json
{
  "THEHELP_TWILIO_KEY": "your AccountSID",
  "THEHELP_TWILIO_TOKEN": "your AuthToken"
}
```

We'll save the configuration for __receiving__ SMS and email for a bit later.


## Sending messages

With that all set up, it's as easy as this:

### Send email

```javascript
var Sendgrid = require('thehelp-messaging').Sendgrid;
var sendgrid = new Sendgrid();

var email = {
  from: 'someone@somewhere',
  to: 'recipient@somewhere',
  subject: 'subject!',

  // this or html is required
  text: 'Plaintext message body!!1',

  // optional
  fromname: 'User Name'
};

sendgrid.send(email, function(err) {
  if (err) {
    throw err;
  }
});
```

[Detailed Sendgrid API documentation](https://sendgrid.com/docs/API_Reference/Web_API/mail.html).

## Send SMS

```javascript
var Twilio = require('thehelp-messaging').Twilio;
var twilio = new Twilio();

var sms = {
  From: '+15551000000',
  To: '+15551000000',
  Body: 'my first text message!'
};

twilio.send(sms, function(err) {
  if (err) {
    throw err;
  }
});
```

Yep, those key names are all capitalized. :0(
[Detailed Twilio API documentation](https://www.twilio.com/docs/api/rest/sending-sms).


## Receiving messages

This project includes some `express` middleware helpers for receiving SMS from Twilio and Email from Sendgrid.

### Receive mail

You'll need one new environment variable:

```json
{
  "THEHELP_SENDGRID_VERIFY": "a hard-to-guess value",
}
```

And you'll need to download and install the [`busboy`](https://www.npmjs.org/package/busboy) node module and (1.2.9 recommended) supply it to the `Sendgrid` class on construction:

```javascript
var express = require('express');
var Busboy = require('busboy');
var Sendgrid = require('thehelp-messaging').Sendgrid;

var app = express();
var sendgrid = new Sendgrid({
  Busboy: Busboy
});

app.post('/sendgrid/email', sendgrid.validate, sendgrid.parse, function(req, res) {
  console.log('email:', req.body);
  res.end();
});
```

[`validate`](http://thehelp.github.io/messaging/src/server/sendgrid.html) will ensure that the message is really from Sendgrid (via the 'verify' querystrying parameter and your environment variable. [`parse`](http://thehelp.github.io/messaging/src/server/sendgrid.html) will use `busboy` to parse out all the non-file components sent by Sendgrid.

Now you just need to set up the [Sengrid Parse dashboard](https://sendgrid.com/developer/reply) to point a given email subdomain of your site to your server. This is kind of a pain to test by deploying to your server all the time, so check out [ngrok](https://ngrok.com/) for exposing a port on your machine to the outside world.

### Receive SMS

Twilio messages are easier to deal with, because they're more easily parsed. However, you still have the problem of ensuring that the message is really from Twilio. That's where `twilio.validate()` comes in - note the new node modules required:

```javascript
var express = require('express');
var bodyPraser = require('body-parser');
var twilioSdk = require('twilio');
var Twilio = require('thehelp-messaging').Twilio;

var app = express();
var twilio = new Twilio({
  twilio: twilioSdk
});

app.post('/twilio/sms', bodyParser.urlencoded(), twilio.validate, function(req, res) {
  console.log('sms:', req.body);
  res.end();
});
```

Now you need to buy a phone number on Twilio and [have it forward SMS sent to it to your application](https://www.twilio.com/user/account/phone-numbers/incoming). Again, check out [ngrok](https://ngrok.com/) for exposing a port on your machine to the outside world. It makes iterating on your SMS setup that much faster.

There are a few additional troublshooting tips in the [`Twilio.validate` detailed documentation.]((http://thehelp.github.io/messaging/src/server/twilio.html)).


## Detailed Documentation

Detailed docs be found at this project's GitHub Pages, thanks to `groc`: <http://thehelp.github.io/messaging>


## Contributing changes

It's a pretty involved project. You'll need Sendgrid and Twilio accounts, and all the environment variables mentioned above.

### Running tests

The unit tests are quick and easy, but the manual tests (not part of the `grunt` 'default' task) in this project are pretty involved. They:

  1) send SMS and email to a phone number and email address for manual verification, and
  2) send SMS and email and then receive those messages programmatically

You'll need some additional environment variables:

```json
{
  "TEST_EMAIL_FROM": "who the emails will be from",
  "TEST_SMS_FROM": "an SMS-capable phone number you own on Twilio",

  "TEST_EMAIL_MANUAL_RECEIVE": "where you'll get emails for manual verify",
  "TEST_SMS_MANUAL_RECEIVE": "where you'll get SMS for manual verify",

  "TEST_EMAIL_RECEIVE": "email set up for receive by Sendgrid",
  "TEST_SMS_RECEIVE": "an SMS-capable phone number you own on Twilio"
}
```

Those last two environment variables are where things get really interesting. You'll need to set up Sendgrid and Twilio to forward messages to your machine at port `3000`. See the 'Receiving messages' section above.

### Pull requests

When you have some changes ready, please include:

* Justification - why is this change worthwhile? Link to issues, use code samples, etc.
* Documentation changes for your code updates. Be sure to check the groc-generated HTML with `grunt doc`
* A description of how you tested the change. Don't forget about the very-useful `npm link` command :0)

I may ask you to use a `git rebase` to ensure that your commits are not interleaved with commits already in the history. And of course, make sure `grunt` completes successfully (take a look at the requirements for [`thehelp-project`](https://github.com/thehelp/project)). :0)


## License

(The MIT License)

Copyright (c) 2013 Scott Nonnenberg &lt;scott@nonnenberg.com&gt;

Permission is hereby granted, free of charge, to any person obtaining
a copy of this software and associated documentation files (the
'Software'), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


History.md
## 1.1.0 (2015-03-22)

* With upgrade of `superagent` to 1.x, the exact set of errors returned by calls has changed. Notable changes: new `status` property, and the `message` is now in this format: `<ENGLISH HTTP CODE, like 'Unauthorized'> - <DETAIL>`
* Add node 0.12 and iojs 1.5/1.6 to travis config
* Update dependencies
* Remove docs from npm package

## 1.0.2 (2014-12-04)

* `Twilio`: properly set `development = true` when `NODE_ENV === 'development'`

## 1.0.1 (2014-10-02)

* Properly handle connectivity/other errors thrown by `superagent` - supply [callback with arity of two](http://visionmedia.github.io/superagent/#error-handling)

## 1.0.0 (2014-09-27)

Breaking changes:

* All environment variables now prefixed with 'THEHELP_'
* Capitalized key names now required for `Twilio.send()`, happily this allows additional paramebers to be included in the request
* `Sendgrid.send()` now requires either `text` or `html` keys be provided. Similarly, means that additional parameters can be add for inclusion in the request to Sendgrid.

Other updates:

* Comprehensive testing: unit testing and manual tests for sending/receiving messages
* All test-related environment variables now prefixed with 'TEST_'
* Removed `nodemailer`, `lodash` and `winston` as dependencies
* Add support for receiving and validating incoming Sendgrid/Twilio messages
* Pare down what's in npm package
* Updated dev dependencies

## 0.1.3 (2014-03-25)

* Minor version updates: `async`, `superagent`
* Patch updates: `nodemailer`
* Updated a few dev dependencies

## 0.1.2 (2014-03-13)

* Patch updates: grunt, thehelp-core, thehelp-test
* Minor version updates: thehelp-project, nodemailer, superagent
* Fixed too-long lines after thehelp-project upgrade

## 0.1.1 (2013-12-18)

* Fixing package.json parse errors

## 0.1.0 (2013-12-14)

* Twilio and Sendgrid are functional


test/start_server.js

var express = require('express');
var bodyParser = require('body-parser');
var morgan = require('morgan');

var app = express();

app.enable('trust proxy');

app.use(morgan('combined'));
app.use(bodyParser.urlencoded({
  extended: true
}));

var port = 3000;
console.log('Starting test server, listening on port ' + port);
app.listen(port);

module.exports = app;


test/unit/server/test_index.js

'use strict';

var test = require('thehelp-test');
var expect = test.expect;

var index = require('../../../src/server/index');

describe('thehelp-twilio', function() {

  it('has property Sendgrid', function() {
    /*jshint -W030 */
    expect(index).to.have.property('Sendgrid').that.exist;
  });

  it('has property Twilio', function() {
    /*jshint -W030 */
    expect(index).to.have.property('Twilio').that.exist;
  });

});


test/unit/server/test_twilio.js

'use strict';

var test = require('thehelp-test');
var expect = test.expect;
var sinon = test.sinon;

var Twilio = require('../../../src/server/twilio');

describe('Twilio', function() {
  var twilio, key, token;

  beforeEach(function() {
    var key, token;

    key = process.env.THEHELP_TWILIO_KEY;
    token = process.env.THEHELP_TWILIO_TOKEN;

    twilio = new Twilio({
      key: 'something',
      token: 'another thing'
    });
  });

  afterEach(function() {
    process.env.THEHELP_TWILIO_KEY = key;
    process.env.THEHELP_TWILIO_TOKEN = token;
  });

  describe('constructor', function() {
    /*jshint nonew: false */

    it('throws if THEHELP_TWILIO_KEY is not defined', function() {
      delete process.env.THEHELP_TWILIO_KEY;

      expect(function() {
        new Twilio();
      }).to['throw']().that.match(/twilio key/);
    });

    it('does not throw if key param is provided', function() {
      delete process.env.THEHELP_TWILIO_KEY;

      expect(function() {
        new Twilio({
          key: 'something'
        });
      }).not.to['throw']();
    });

    it('throws if THEHELP_TWILIO_TOKEN is not defined', function() {
      delete process.env.THEHELP_TWILIO_TOKEN;

      expect(function() {
        new Twilio();
      }).to['throw']().that.match(/twilio token/);
    });

    it('does not throw if token param is provided', function() {
      delete process.env.THEHELP_TWILIO_TOKEN;

      expect(function() {
        new Twilio({
          token: 'something'
        });
      }).not.to['throw']();
    });
  });

  // Sending SMS

  describe('#send', function() {
    it('returns error if To is not provided', function(done) {
      var email = {
        From: 'from',
        Body: 'body'
      };

      twilio.send(email, function(err) {
        expect(err).to.have.property('message').that.match(/options.To/);
        done();
      });
    });

    it('returns error if From is not provided', function(done) {
      var email = {
        To: 'to',
        Body: 'body'
      };

      twilio.send(email, function(err) {
        expect(err).to.have.property('message').that.match(/options.From/);
        done();
      });
    });

    it('returns error if Body is not provided', function(done) {
      var email = {
        To: 'to',
        From: 'from'
      };

      twilio.send(email, function(err) {
        expect(err).to.have.property('message').that.match(/options.Body/);
        done();
      });
    });
  });

  describe('#_sendFinish', function() {
    it('returns error if error is provided, null res', function(done) {
      var expected = new Error('superagent error');
      twilio._sendFinish(expected, null, null, function(err) {
        expect(err).to.have.property('message', expected.message);

        done();
      });
    });

    it('updates error message if body.detail', function(done) {
      var response = 'error from twilio';

      var err = new Error('original');
      var res = {
        status: 400,
        body: {
          detail: response
        }
      };
      var options = {
        To: 'blah',
        From: 'blah',
        Body: 'something'
      };

      twilio._sendFinish(err, res, options, function(err) {
        expect(err).to.have.property('message').that.equal('original - ' + response);
        expect(err).to.have.property('options').that.deep.equal(options);
        expect(err).not.to.have.property('response');

        done();
      });
    });
  });

  // SMS Processing

  describe('#truncate', function() {
    it('takes string down to 70 characters if it has unicode', function() {
      var text = '©2345678901234567890123456789012345678901234567890' + // 50 chars
             '123456789012345678901'; // 21 chars
      expect(text).to.have.length(71);
      var expected = '©2345678901234567890123456789012345678901234567890' + // 50 chars
                 '12345678901234567...'; // 20 chars

      var actual = twilio.truncate(text);
      expect(actual).to.equal(expected);
    });

    it('takes string down to 159 chars if it has an escape character', function() {
      var text = '[2345678901234567890123456789012345678901234567890' + // 50 chars
             '12345678901234567890123456789012345678901234567890' + // 50 chars
             '12345678901234567890123456789012345678901234567890' + // 50 chars
             '1234567890'; // 10 chars
      expect(text).to.have.length(160);
      var expected = '[2345678901234567890123456789012345678901234567890' + // 50 chars
                 '12345678901234567890123456789012345678901234567890' + // 50 chars
                 '12345678901234567890123456789012345678901234567890' + // 50 chars
                 '123456...'; // 9 chars
      var actual = twilio.truncate(text);
      expect(actual).to.equal(expected);
    });

    it('takes string down to 140 chars buffer is provided', function() {
      var text = '12345678901234567890123456789012345678901234567890' + // 50 chars
             '12345678901234567890123456789012345678901234567890' + // 50 chars
             '12345678901234567890123456789012345678901234567890' + // 50 chars
             '123456789012345678901234567890'; // 30 chars
      expect(text).to.have.length(180);
      var expected = '12345678901234567890123456789012345678901234567890' + // 50 chars
                 '12345678901234567890123456789012345678901234567890' + // 50 chars
                 '1234567890123456789012345678901234567...'; // 40 chars
      var actual = twilio.truncate(text, 20);
      expect(actual).to.equal(expected);
    });
  });

  describe('#escapeCharacterCount', function() {
    it('return zero for normal charaters', function() {
      expect(twilio.escapeCharacterCount('abcde')).to.equal(0);
    });

    it('handles all characters requiring an escape character', function() {
      expect(twilio.escapeCharacterCount('|^{}€[~]\\')).to.equal(9);
    });
  });

  describe('#containsUnicode', function() {
    it('returns false for ASCII', function() {
      expect(twilio.containsUnicode('abcde')).to.equal(false);
    });

    it('returns true for non-ASCII', function() {
      expect(twilio.containsUnicode('©')).to.equal(true);
    });
  });

  describe('#_truncate', function() {
    it('handles strings right at the limit', function() {
      expect(twilio._truncate(5, 'abcde')).to.equal('abcde');
    });

    it('handles strings smaller than the limit', function() {
      expect(twilio._truncate(5, 'abcd')).to.equal('abcd');
    });

    it('strings over the limit', function() {
      expect(twilio._truncate(5, 'abcdefgh')).to.equal('ab...');
    });

    it('strings just a little over the limit', function() {
      expect(twilio._truncate(5, 'abcdef')).to.equal('ab...');
    });
  });

  // Express middleware

  describe('#validate', function() {
    it('returns an error if this.twilio is not set', function(done) {
      delete twilio.twilio;

      twilio.validate({}, null, function(err) {
        expect(err).to.have.property('message').that.match(/options.twilio/);
        done();
      });
    });

    it('returns error if twilio.validateExpressRequest returns false', function(done) {
      twilio.twilio = {
        validateExpressRequest: sinon.stub().returns(false)
      };
      var req = {
        body: {
          From: 'from',
          To: 'to'
        }
      };

      twilio.validate(req, null, function(err) {

        expect(twilio).to.have
          .deep.property('twilio.validateExpressRequest.callCount', 1);

        expect(err).to.have.property('message').that.match(/not pass twilio validation/);
        expect(err).to.have.property('body').that.deep.equal(req.body);
        expect(err).to.have.property('status', 400);
        expect(err).to.have.property('text').that.match(/twilio, are you?/);

        done();
      });
    });

    it('does not call validateExpressRequest if development set to true', function(done) {
      twilio.twilio = {
        validateExpressRequest: sinon.stub().returns(false)
      };
      twilio.development = true;

      twilio.validate({}, null, function(err) {

        expect(twilio).to.have
          .deep.property('twilio.validateExpressRequest.callCount', 0);

        expect(err).not.to.exist;

        done();
      });
    });
  });

});


src/server/sendgrid.js
/*
# Sendgrid
This class is a thin wrappper around the Sendgrid Send mail REST endpoint:
<https://sendgrid.com/docs/API_Reference/Web_API/mail.html>.
*/

'use strict';

var superagent = require('superagent');

/*
Two environment varialbes are required to create a `Sendgrid` instance:

1. `username`/`THEHELP_SENDGRID_USERNAME` - this is your raw Sendgrid account username
2. `password`/`THEHELP_SENDGRID_PASSWORD` - unfortunately, this is the same password as
your account. Not an account API access token. I heartily recommend that you create
additional sub-accounts to use for your programs:
<https://sendgrid.com/docs/User_Guide/multiple_credentials.html>.
*/
function SendGrid(options) {
  /*jshint maxcomplexity: 8 */

  options = options || {};

  this.username = options.username || process.env.THEHELP_SENDGRID_USERNAME;
  if (!this.username) {
    throw new Error('need to provide sendgrid username by options or env var');
  }

  this.password = options.password || process.env.THEHELP_SENDGRID_PASSWORD;
  if (!this.password) {
    throw new Error('need to provide sendgrid password by options or env var');
  }

  this.verify = options.verify || process.env.THEHELP_SENDGRID_VERIFY;
  this.Busboy = options.Busboy;

  this.parse = this.parse.bind(this);
  this.validate = this.validate.bind(this);

  this.superagent = options.superagent || superagent;
}

// Commonly-Used Methods
// --------

/*
`send` checks that all four required keys are included on the provided
`options` object:

+ `to`
+ `from`,
+ `text` or `html`
+ `subject`

More documentation on the endpoint we hit:
<https://sendgrid.com/docs/API_Reference/Web_API/mail.html>. We pass your
payload directly to it, after adding your sendgrid credentials.
*/
SendGrid.prototype.send = function send(options, cb) {
  /*jshint maxcomplexity: 8 */
  /*jshint camelcase: false */
  var _this = this;

  options = options || {};

  if (!options.to) {
    return cb(new Error('sendgrid/send: need options.to!'));
  }
  if (!options.from) {
    return cb(new Error('sendgrid/send: need options.from!'));
  }
  if (!options.text && !options.html) {
    return cb(new Error('sendgrid/send: need either options.text or options.html!'));
  }
  if (!options.subject) {
    return cb(new Error('sendgrid/send: need options.subject!'));
  }

  options.api_user = options.api_user || this.username;
  options.api_key = options.api_key || this.password;

  this.superagent
    .post('https://api.sendgrid.com/api/mail.send.json')
    .type('form')
    .send(options)
    .end(function(err, res) {
      _this._sendFinish(err, res, options, cb);
    });
};

// `_sendFinish` handles the payload returned to us from the call to Sendgrid.
SendGrid.prototype._sendFinish = function _sendFinish(err, res, options, cb) {
  if (err) {
    res = res || {};
    if (res.body && res.body.errors && res.body.errors.length) {
      err.message += ' - ' + res.body.errors[0];

      if (res.body.errors.length > 1) {
        err.errors = res.body.errors;
      }
      delete err.response;
    }
    err.options = options;
    return cb(err);
  }

  return cb(null, res.body);
};

// Express middleware
// =======

/*
To help you deal with requests coming from Sendgrid, described here:
<https://sendgrid.com/docs/API_Reference/Webhooks/parse.html>

_Note: Don't forget to disable CSRF checking for your incoming options endpoint!_
*/

/*
It's kind of a pain to parse incoming Sendgrid optionss because their content type is
'multipart/form-data'. This middleware function pulls all non-file data out of the
Sendgrid request, and then makes that available at `req.body`.

Use it like this:
```
app.post('/sendgrid/options', sendgrid.parse, function(req, res) {
  console.log(req.body)
  res.status(200);
  res.end();
});
```

_Note: You'll need to supply [`Busboy`](https://github.com/mscdex/busboy) to this class
on construction if you want to use this middleware method._
*/
SendGrid.prototype.parse = function parse(req, res, next) {
  var type = req.headers['content-type'] || '';
  req.body = req.body || {};

  if (!this.Busboy) {
    var err = new Error('Need to set options.Busboy on construction!');
    return next(err);
  }

  if (type.indexOf('multipart') < 0) {
    return next();
  }

  var busboy = new this.Busboy({headers: req.headers});

  busboy.on('field', function(name, value) {
    req.body[name] = value;
  });

  busboy.on('finish', function() {
    next();
  });

  req.pipe(busboy);
};


/*
But we don't want to accept incoming emails from just anyone. We want to be sure we're
really dealing with Sendgrid. On the
[inbound dashboard](<https://sendgrid.com/developer/reply>) you can add a querystring to
the URL Sendgrid will hit.

This method assumes that you've set up a `verify` querystring and the correct value of
that querystring is either set in the `SENDGRID_VERIFY` environment variable or
directly passed to this class on construction.

We just add one new middleware function:

```
app.post('/sendgrid/options', sendgrid.validate, sendgrid.parse, function(req, res) {
  console.log(req.body)
  res.status(200);
  res.end();
})
```
*/
SendGrid.prototype.validate = function validate(req, res, next) {
  var err;

  if (!this.verify) {
    err = new Error(
      'Need to provide sendgrid verify value on construction or via env var'
    );
    return next(err);
  }

  if (req.query.verify !== this.verify) {
    err = new Error('Request did not pass sendgrid validation');
    err.body = req.body;
    err.query = req.query;
    err.status = 400;
    err.text = 'You aren\'t sendgrid, are you?';
    return next(err);
  }

  next();
};


module.exports = SendGrid;


src/server/twilio.js
/*
# Twilio
This class makes it easy to send text messages with
[Twilio's REST API](http://www.twilio.com/docs/api/rest). Two environment variables
are required:

1. TWILIO_KEY - your "Account SID" on your
[Account Detail](https://www.twilio.com/user/account) page.
2. TWILIO_TOKEN - your "Auth Token" on that same page.
*/

'use strict';

var superagent = require('superagent');

function Twilio(options) {
  /*jshint maxcomplexity: 9 */

  options = options || {};

  this.key = options.key || process.env.THEHELP_TWILIO_KEY;
  if (!this.key) {
    throw new Error('need to provide twilio key by options or env var!');
  }

  this.token = options.token || process.env.THEHELP_TWILIO_TOKEN;
  if (!this.token) {
    throw new Error('need to provide twilio token by options or env var!');
  }

  this.twilio = options.twilio;
  this.development = options.development;
  if (typeof this.development === 'undefined') {
    this.development = process.env.NODE_ENV === 'development';
  }

  this.validate = this.validate.bind(this);

  this.superagent = options.superagent || superagent;
}

// Sending SMS
// --------

/*
`send` assembles the right request for
[Twilio's Send REST API](https://www.twilio.com/docs/api/rest/sending-sms),
parsing the return value for successful send status and returning an error
if it isn't found.

_NOTE: It will truncate any message you send if necessary, warning in the log when
that becomes necessary._
*/
Twilio.prototype.send = function send(options, cb) {
  var _this = this;
  options = options || {};

  if (!options.To) {
    return cb(new Error('twilio/send: need options.To! (capitalized)'));
  }
  if (!options.From) {
    return cb(new Error('twilio/send: need options.From! (capitalized)'));
  }
  if (!options.Body) {
    return cb(new Error('twilio/send: need options.Body! (capitalized)'));
  }

  this.superagent
    .post('https://api.twilio.com/2010-04-01/Accounts/' + this.key + '/SMS/Messages.json')
    .auth(this.key, this.token)
    .type('form')
    .send(options)
    .end(function(err, res) {
      _this._sendFinish(err, res, options, cb);
    });
};

// `_sendFinish` handles the payload returned to us from the call to Twilio.
Twilio.prototype._sendFinish = function _sendFinish(err, res, options, cb) {
  if (err) {
    res = res || {};
    if (res.body && res.body.detail) {
      err.message += ' - ' + res.body.detail;
      delete err.response;
    }
    err.options = options;
    return cb(err);
  }

  return cb(null, res.body);
};

// SMS processing
// --------

/*
`getMaxLength` goes through the full set of steps required to capture the maximumn
length for an SMS, so it doesn't get truncated on send:

1. The `max` starts at 160
2. If the string contains unicode, that `max` goes down to 70
3. If there's some additional text that still needs to be added to the message, of length
`buffer`, we subtract that from the `max`.
4. We subract the number of characters requiring escaping from the `max`.
*/
Twilio.prototype.getMaxLength = function getMaxLength(text, buffer) {
  var max = 160;

  if (this.containsUnicode(text)) {
    max = 70;
  }

  if (buffer) {
    max -= buffer;
  }

  max -= this.escapeCharacterCount(text);

  return max;
};

/*
`truncate` calculates the max length for the provided `text`, and then does the
truncation for you. If a truncation is required, the last three characters will be '...'.
*/
Twilio.prototype.truncate = function truncate(text, buffer) {
  var max = this.getMaxLength(text, buffer);
  return this._truncate(max, text);
};

// `escapeCharacterCount` finds the number of characters that require escaping
// io text messages, as listed in this
// [blog post](http://www.twilio.com/engineering/2012/11/08/adventures-in-unicode-sms).
Twilio.prototype.escapeCharacterCount = function escapeCharacterCount(text) {
  var result = 0;
  var escapes = /[\|\^{}€\[~\]\\]/g;
  var match = text.match(escapes);
  if (match) {
    result += match.length;
  }

  return result;
};

/*
`containsUnicode` tries to determine if a text contains characters which will flip the
text message infrastructure into a different encoding, reducing the number of
available characters to 70.

We do a simple check, looking for just the first block of characters from
[Part 1: Latin alphabet #1 (ISO/IEC 8859-1)](http://en.wikipedia.org/wiki/ISO/IEC_8859-1).
It's an imperfect stand-in for the characters actually supported
in the [GSM 7-bit encoding system](http://bit.ly/1lCFVcJ), so if texting ever gets very
important to an app, and it's sending input from users, we'll need to get a lot better at
this.
*/
Twilio.prototype.containsUnicode = function containsUnicode(text) {
  return (/[^\u0000-\u007E]/).test(text);
};

// `truncate` returns a string with `limit` characters or less. If the original string
// was longer than `limit` characters, it will be truncated to fit. Any truncated
// string will end with an ellipsis ("...") to signify that it's missing info.
Twilio.prototype._truncate = function _truncate(limit, text) {
  var result;
  if (text.length > limit) {
    result = text.substring(0, limit - 3);
    result += '...';
    return result;
  }
  return text;
};

// Express middleware
// =======

/*
Documentation kind of sucks for incoming SMS on Twilio, but there is some:

+ <https://www.twilio.com/docs/quickstart/php/sms/hello-monkey>
*/

/*
It just so happens that the `twilio` npm package provides a
[nice little function](http://twilio.github.io/twilio-node/#validateExpressRequest) to
validate an incoming request. This little bit of middleware calls that when not in
development mode, using the Twilio API Key and Token you've already provided.

Just put it in front of your POST handler:
```
app.post('/twilio/sms', twilio.validate, function(req, res) {
  console.log(req.body);
  res.status(200);
  res.end();
})
```

Some trouble-shooting tips:

+ To use this, you have to install the `twilio` node module and pass it in on construction
+ Did you install any middleware to parse the
[`urlencoded`](https://github.com/expressjs/body-parser#bodyparserurlencodedoptions)
body of the request?
+ Disable CSRF checking for this endpoint
+ The validation function gets the `host` and `protocol` from Express. If you're behind
a proxy, it's easiest to `app.enable('trust proxy');`
+ As of August 2014, Twilio does not support TLS. Your https server needs to support some
 version of SSL.
 [More information on how to set up SSL securely](https://www.ssllabs.com/index.html).

*/
Twilio.prototype.validate = function validate(req, res, next) {
  var err;

  if (!this.twilio) {
    err = new Error('Need to set options.twilio');
    return next(err);
  }

  if (!this.development && !this.twilio.validateExpressRequest(req, this.token)) {
    err = new Error('Request did not pass twilio validation');
    err.body = req.body;
    err.status = 400;
    err.text = 'You aren\'t twilio, are you?';
    return next(err);
  }

  next();
};


module.exports = Twilio;
